### ISSUE TITLE
Hashing Vector2i often results in a hash collision

### ISSUE THREAD

**Issue by michael-oti** [godotengine/godot#51023]
### Godot version

v4.0.dev.custom_build [60eb508fb]

### System information

Arch Linux

### Issue description

Hashing Vector2i often results in a hash collision. Does not happen with Vector2.
For example:
`print(hash(Vector2i(0, 33)))` prints `5859942`
`print(hash(Vector2i(1, 0)))` prints `5859942`

### Steps to reproduce

```
extends Node2D

var dict := {}

func _ready():

	for x in 100:
		for y in 100:
			var vec2i := Vector2i(x, y)
			var vec2i_hash := hash(vec2i)
			if dict.has(vec2i_hash):
				dict[vec2i_hash].append(vec2i) 
			else:
				dict[vec2i_hash] = [vec2i]
	
	for key in dict:
		var value: Array = dict[key]
		if len(value) > 1 :
			print('Vector2i %s have same hash() -> %s' % [value, key])

```

### Minimal reproduction project

[hash_vector2i.zip](https://github.com/godotengine/godot/files/6901855/hash_vector2i.zip)

**Comment by logzero** [comment]
Godot uses djb2:
`hash = ((hash << 5) + hash) + c; /* hash * 33 + c */`

So (0, n*33) will collide with (n, 0)

A fix would be use djb2a which doesn't have this issue:
`hash = ((hash << 5) + hash) ^ c; /* hash * 33 ^ c */`

I prefer FNV-1a, as it seems to have a better random distribution. But that is subjective.

**Comment by Calinou** [comment]
Changing the vector hash may break compatibility with existing projects in subtle ways. I suppose we can still do this in 3.4, we have to duly note it in the changelog.

**Comment by Calinou** [comment]
> A fix would be use djb2a which doesn't have this issue:
> `hash = ((hash << 5) + hash) ^ c; /* hash * 33 ^ c */`

Feel free to open a pull request to use djb2a :slightly_smiling_face:

**Comment by kleonc** [comment]
> ### Steps to reproduce
> 
> ```
> extends Node2D
> 
> var dict := {}
> 
> func _ready():
> 
> 	for x in 100:
> 		for y in 100:
> 			var vec2i := Vector2i(x, y)
> 			var vec2i_hash := hash(vec2i)
> 			if dict.has(vec2i_hash):
> 				dict[vec2i_hash].append(vec2i) 
> 			else:
> 				dict[vec2i_hash] = [vec2i]
> 	
> 	for key in dict:
> 		var value: Array = dict[key]
> 		if len(value) > 1 :
> 			print('Vector2i %s have same hash() -> %s' % [value, key])
> ```

In v4.0.beta4.official [e6751549c] it prints nothing. Most likely fixed by #61934. Hence closing.
